src/cpp/shared/game/game.cpp

```diff
@@ -268,130 +267,124 @@ void game::ProcessOnMapHeroes() {
   hero *randomHero;
   char mightBeHeroAlreadyExists[56];
   int ppMapExtraHeroIdx;
-  heroMapExtra *mapExtraHero;
+  HeroExtra *mapExtraHero;
   char isJail;
   int i;
 
-  result = memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
-  for (i = 0; i < 3; ++i) {
-    for (coordYForRandomHero = 0; ; ++coordYForRandomHero) {
-      result = (void *)MAP_HEIGHT;
-      if (coordYForRandomHero >= MAP_HEIGHT)
-        break;
-      for (coordXForRandomHero = 0; coordXForRandomHero < MAP_WIDTH; ++coordXForRandomHero) {
-        loc = &this->map.tiles[(coordYForRandomHero * this->map.width) + coordXForRandomHero];
-        if ((loc->objType & 0x7F) == 55 || loc->objType == 251) {
-          isJail = (loc->objType & 0x7F) == LOCATION_JAIL;
-          ppMapExtraHeroIdx = loc->extraInfo;
-          mapExtraHero = (heroMapExtra *)ppMapExtra[ppMapExtraHeroIdx];
-
-          if (!i) {                           // 1st iteration
-            if (!mapExtraHero->field_11 || mapExtraHero->heroID >= 54 || mightBeHeroAlreadyExists[mapExtraHero->heroID]) {
-              mapExtraHero->couldBeHasFaction = 0;
-            } else {
-              mightBeHeroAlreadyExists[mapExtraHero->heroID] = 1;
-              mapExtraHero->couldBeHasFaction = 1;
-            }
-            if (isJail) {
-              mapExtraHero->owner = -1;
-            } else {
-              mapExtraHero->owner = loc->objectIndex / 7; // Uh, redundant line?
-              mapExtraHero->owner = gcColorToPlayerPos[mapExtraHero->owner];
-            }
+  memset(mightBeHeroAlreadyExists, 0, 54u);// Only 54 out of 56 bytes set to zero, for whatever reason.
+  for (coordYForRandomHero = 0; coordYForRandomHero < MAP_HEIGHT; ++coordYForRandomHero) {
+    for (coordXForRandomHero = 0; coordXForRandomHero < MAP_WIDTH; ++coordXForRandomHero) {
+      loc = &this->map.tiles[(coordYForRandomHero * this->map.width) + coordXForRandomHero];
+      if ((loc->objType & 0x7F) == 55 || loc->objType == 251) {
+        isJail = (loc->objType & 0x7F) == LOCATION_JAIL;
+        ppMapExtraHeroIdx = loc->extraInfo;
+        mapExtraHero = (HeroExtra *)ppMapExtra[ppMapExtraHeroIdx];
+        if (!mapExtraHero->field_11 || mapExtraHero->heroID >= 54 || mightBeHeroAlreadyExists[mapExtraHero->heroID]) {
+          mapExtraHero->couldBeHasFaction = 0;
+        } else {
+          mightBeHeroAlreadyExists[mapExtraHero->heroID] = 1;
+          mapExtraHero->couldBeHasFaction = 1;
+        }
+
+        if (isJail) {
+          mapExtraHero->owner = -1;
+          faction = mapExtraHero->factionID;
+        } else {
+          mapExtraHero->owner = gcColorToPlayerPos[loc->objectIndex / 7];
+          faction = loc->objectIndex % 7; // Constant here (needs to be generalized)
+          if (faction == FACTION_MULTIPLE) { // Constant here (faction related)
```

Comment:
I think this is actually representing a random faction, not multiple factions. Yes, the game uses the constant 6 in a faction-related context to mean different things in different places. In a third context, it also represents the Captain for the Knight faction. This is something that's going to cause trouble for us.

Reply:
Yeah. I'm not sure how this will be addressed yet. I'm leaving this here for now, as I'm currently not working on the generalization yet.